velox/functions/lib/Re2Functions.h

```diff
@@ -30,32 +30,95 @@ namespace facebook::velox::functions {
+  /// kRelaxedFixed preceded by one or more '%', such as '%_pr_es_to_',
+  /// '%%%_pr_es_to_'.
+  kRelaxedSuffix,
   /// Patterns matching '%{c0}%', such as '%foo%%', '%%%hello%'.
   kSubstring,
   /// Patterns which do not fit any of the above types, such as 'hello_world',
   /// '_presto%'.
   kGeneric,
 };
 
-struct PatternMetadata {
-  PatternKind patternKind;
-  // Contains the length of the unescaped fixed pattern for patterns of kind
-  // kFixed, kPrefix, kSuffix and kSubstring. Contains the count of wildcard
-  // character '_' for patterns of kind kExactlyN and kAtLeastN. Contains 0
-  // otherwise.
-  size_t length;
-  // Contains the unescaped fixed pattern in patterns of kind kFixed, kPrefix,
-  // kSuffix and kSubstring.
-  std::string fixedPattern = "";
+struct SubPatternMetadata {
+  enum class SubPatternKind {
+    kSingleCharWildcard,
+    kAnyCharsWildcard,
+    kLiteralString
+  };
+
+  SubPatternKind kind;
+  /// Pattern string of the kind.
+  /// For kSingleCharWildcard, it could be '___';
+  /// For kAnyCharsWildcard, it could be '%%';
+  /// For kLiteralString, it could be 'abc'.
+  std::string pattern;
+};
+
+class PatternMetadata {
+ public:
+  static PatternMetadata generic();
+  static PatternMetadata atLeastN(size_t length);
+  static PatternMetadata exactlyN(size_t length);
+  static PatternMetadata fixed(std::vector<SubPatternMetadata> subPatterns);
+  static PatternMetadata relaxedFixed(
+      std::vector<SubPatternMetadata> subPatterns);
+  static PatternMetadata prefix(std::vector<SubPatternMetadata> subPatterns);
+  static PatternMetadata relaxedPrefix(
+      std::vector<SubPatternMetadata> subPatterns);
+  static PatternMetadata suffix(std::vector<SubPatternMetadata> subPatterns);
+  static PatternMetadata relaxedSuffix(
+      std::vector<SubPatternMetadata> subPatterns);
+  static PatternMetadata substring(std::vector<SubPatternMetadata> subPatterns);
+
+  const PatternKind patternKind() const {
+    return patternKind_;
+  }
+
+  const size_t length() const {
+    return length_;
+  }
+
+  const std::vector<SubPatternMetadata>& subPatterns() const {
+    return subPatterns_;
+  }
+
+  const std::string& fixedPattern() const {
+    return fixedPattern_;
+  }
+
+ private:
+  PatternMetadata(
+      PatternKind patternKind,
+      size_t length,
+      std::vector<SubPatternMetadata> subPatterns);
+
+  PatternKind patternKind_;
+  /// Contains the length of the unescaped fixed pattern for patterns of kind
+  /// k[Relaxed]Fixed, k[Relaxed]Prefix, k[Relaxed]Suffix and
+  /// k[Relaxed]Substring. Contains the count of wildcard character '_' for
+  /// patterns of kind kExactlyN and kAtLeastN. Contains 0/// otherwise.
+  size_t length_;
+  /// Contains the fixed pattern in patterns of kind k[Relaxed]Fixed,
+  /// k[Relaxed]Prefix, k[Relaxed]Suffix and k[Relaxed]Substring.
+  std::string fixedPattern_;
+  /// Contains the sub-patterns for k[Relaxed]Xxx patterns.
+  std::vector<SubPatternMetadata> subPatterns_;
```

Comment:
This works, but I wonder if a simpler way to model this would be to have a single string (_h_e_ll_o_) and a vector of pairs of start + length to indicate a list of substrings to match:

_h_e_ll_o_
[[1, 1], [3, 1], [5, 2], [7, 1]]

This way we do not need to create lots of small strings. 

I noticed that BM shows that relaxes patterns a bit slower than fixed. Do you happen to know where the slowness comes from?

Reply:
@mbasmanova Simplied the modeling as you suggested.

For the performance diff, I think its due to the complexity of relaxed pattern, take kPrefix and kRelaxedPrefix as example:

- For kPrefix, for every input row, we only need to compare the prefix string once.
- For kRelaxedPrefix, we need to access `kSubPatternKinds`, `kSubPatternRanges` and do more string matching.

Might need to generate a flamegraph to confirm.